### PR TITLE
HPCC-25470 Add option to avoid scanning all KJ parts when mapping

### PR DESCRIPTION
**Pull request by jakesmith**
HINT(keyedJoinAssumePrimary(true))

Will prevent KJ from checking every physical part in an index
involved in a keyed join.
This can be very expensive if it's a large super index.

When this option is used, it will assume the primary parts
exist, and will not utilize any secondary copies.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please take a look - leaving as draft PR for now to discuss.

This is essentially a workaround for when a very large super index is used.
The current code (without this option on), can take minutes to scan the 1000's of parts involved.

The scan for other copies (if primary is missing) isn't really correct here anyway, it would spot if primary is missing at initialization time, and use replica, but can't spot it once serialized to slave.

**Review by richardkchapman — DISMISSED** [review]

File: thorlcr/thorutil/thormisc.hpp [hunk 1]
```diff
@@ -102,6 +102,7 @@
 #define THOROPT_MAXLFN_BLOCKTIME_MINS "maxLfnBlockTimeMins"     // max time permitted to be blocked on a DFS logical file operation.
 #define THOROPT_VALIDATE_FILE_TYPE    "validateFileType"        // validate file type compatibility, e.g. if on fire error if XML reading CSV    (default = true)
 #define THOROPT_MIN_REMOTE_CQ_INDEX_SIZE_MB "minRemoteCQIndexSizeMb" // minimum size of index file to enable server side handling                (default = 0, meaning use heuristic to determin)
+#define THOROPT_KJ_ASSUME_PRIMARY "keyedJoinAssumePrimary"      // assume priamry part exists (don't check when mapping, which can be slow)
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply thorlcr/thorutil/thormisc.hpp]: Corrected

**Comment by jakesmith** [comment]
In addition to the option, I noticed that a file part was being needlessly rechecked (file->exists()), if the node group contained multiple matches, which will happen if slavesPerNode>1.
I have moved that part exists check out of that loop and avoided the loop completely in the file part doesn't exist.
So that change should speed things up in the normal case (without option) where running on clusters with multiple slaves per node.

@richardkchapman @mckellyln - please review

**Comment by mckellyln** [comment]
Moving the exists() is good because I have seen when that gets called many times rapidly and can cause port exhaustion from TIME_WAITs.

**Comment by richardkchapman** [comment]
@jakesmith Looks like it needs squashing?

**Comment by jakesmith** [comment]
@richardkchapman @mckellyln - squashed.

**Review by mckellyln — APPROVED** [review]
Looks fine.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**